src/features/api/lib/network-provider.ts

```diff
@@ -1,13 +1,19 @@
+type NetworkProviderContext = {
+  client: NetworkProvider
+}
+
 type BeforeRequestInterceptor = (
   request: Request,
-  options: NetworkProviderOptions
-) => Request
+  options: NetworkProviderOptions,
+  context: NetworkProviderContext
+) => Request | Promise<Request>
 
 type AfterRequestInterceptor = (
   request: Request,
   options: NetworkProviderOptions,
-  response: Response
-) => Response
+  response: Response,
+  context: NetworkProviderContext
+) => Response | Promise<Response>
```

Comment:
We are adding a "context" object to all interceptors containing, currently, a reference to the client itself. This could be achieved by managing the `this` context of functions, but in general it's nice to provide a simpler way to reach this object and allow interceptors to be simple arrow functions, simplifying the testability and pureness of these.